### ISSUE TITLE
result.css should probably be a UTF-8-encoded buffer, not a string

### ISSUE THREAD

**Issue by joliss** [sass/node-sass#711]
When non-ASCII characters occur in the output, libsass helpfully emits a `@charset` at-rule. For example,

```sass
p::before { content: "Hëllö world!"; }
```

generates

```css
@charset "UTF-8";
p::before {
  content: "Hëllö world!"; }
```

However, node-sass's `result.css` is a string (a sequence of Unicode characters), so it is up the user of node-sass to decide the encoding. UTF-8 is a common default, but if the user decided to a different encoding (like latin1 or UTF-16), the `@charset` at-rule would be wrong and the file would stop working.

```js
fs.writeFileSync('out.css', result.css, { encoding: 'utf16le' }) // override Node's default of 'utf8'
```

So because we're explicitly declaring that the file is UTF-8, we should actually go ahead and convert it to a Buffer (a sequence of bytes) using the UTF-8 encoding. In other words, what I'm suggesting node-sass should do boils down to this:

```js
function renderSync(...) {
  ...
  res.css = new Buffer(res.css, 'utf8') // encode (this is a hack, of course)
  return res
}
```

Regarding `result.map`: The [source map spec](https://docs.google.com/document/d/1U1RGAehQwRypUTovF1KRlpiOFze0b-_2gc6fAH0KY0k/edit?pli=1) (is this up-to-date?) says: "For simplicity, the character set encoding is always UTF-8." So it might make sense to make `result.map` a UTF-8-encoded buffer as well.

**Comment by am11** [comment]
@joliss, thanks for sharing your thoughts. However, IMVHO it is node-sass API's implementer responsibility to make sure they are using correct `utf8` encoding.

The default encoding of Buffer is `utf8`. The consumer can instantiate a Buffer if they want even with a slimmer version: 

```javascript
var sass = require('node-sass');
var result = sass.renderSync({data: '.test { display: none }' });
var myPreferredCssObj = new Buffer(result.css);  // utf8 encoded Buffer
var myPreferredMapObj = new Buffer(result.map);  // utf8 encoded Buffer
``` 

The reason why we decided not to write output files in first place is because we realized that saving to file-system is not the only scenario out there. Consumers might want the output directly to some stream or store in DB or send it via web service etc. 

Besides, the return values are instantiated via v8 in C++ code and we try our best to forward the raw output back to user without altering it. 

Our aim is to make node-sass layer reflection of libsass. Since libsass returns *stringly-typed* stuff, we tend to forward it to the downstream as is without warranting or promising the authenticity of output. IMO this is the core purpose of node-sass to act like a thin and transparent wrapper around blazingly-fast libsass, as opposed to being opinionated and yet getting thicker by minute, dealing with all kinds of heterogeneous scenarios (defensive programming in node-sass === anti-libsass-performance for all consumers).

**Comment by joliss** [comment]
> Our aim is to make node-sass layer reflection of libsass.

Yes, this is very reasonable and exactly what I expect! :)

I may be slightly off here, but looking at the [`sass_context_get_output_string` signature](https://github.com/sass/libsass/blob/d35892aa574c6592e20575b8c86ec63c5bfe873b/sass_context.h#L97), it seems that libsass returns its output as a `char` array (a sequence of *bytes*, not Unicode characters) that just so happens to be encoded in UTF-8. A `char` array in C would be equivalent to Node's [Buffer object](http://nodejs.org/api/buffer.html#buffer_buffer) rather than a JavaScript string (since a JavaScript string is a sequence of Unicode characters).

Node-sass's current behavior seems to accidentally add another layer of decoding: Rather than passing the `char` sequence it receives from libsass through as a Buffer, it decodes them implicitly into a string of Unicode characters: [binding.cpp:206](https://github.com/sass/node-sass/blob/9cc9b3152fcea9ffc7bbb3b6030a2d6726a9ccaa/src/binding.cpp#L206) calls `NanNew<String>`, which I'm guessing uses V8's [`String::New(const char *)`](http://izs.me/v8-docs/classv8_1_1String.html#af0f9c44d85056bd575c01e8be7cc1b01) function to implicitly decode UTF-8.

So it seems that in order to make node-sass simply pass through the output from libsass (rather than do extra decoding on its own), [binding.cpp:206](https://github.com/sass/node-sass/blob/9cc9b3152fcea9ffc7bbb3b6030a2d6726a9ccaa/src/binding.cpp#L206) might need to be somehow changed to create a Node `Buffer` rather than a `v8::String`.

Well. I hope this is useful! :) If I've failed to convince you, or you want to keep the current behavior for whatever other reason, feel free to just close this.

**Comment by am11** [comment]
> Well. I hope this is useful!

Indeed! :)

> or you want to keep the current behavior for whatever other reason

There is in fact no other reason than keeping it simple. :)

Anyway, here is (one way) to implement it using [`NanNewBufferHandle`](https://github.com/rvagg/nan#localobject-nannewbufferhandlechar--uint32_t-localobject-nannewbufferhandleuint32_t):

```diff
diff --git a/src/binding.cpp b/src/binding.cpp
index 87ffae8..fb93866 100644
--- a/src/binding.cpp
+++ b/src/binding.cpp
@@ -202,7 +202,8 @@ int get_result(sass_context_wrapper* ctx_w, Sass_Context* ctx, bool is_sync = fa
   int status = sass_context_get_error_status(ctx);

   if (status == 0) {
-    NanNew(ctx_w->result)->Set(NanNew("css"), NanNew<String>(sass_context_get_output_string(ctx)));
+    const char* css = sass_context_get_output_string(ctx);
+    NanNew(ctx_w->result)->Set(NanNew("css"), NanNewBufferHandle(css, strlen(css)));
     get_stats(ctx_w, ctx);
     get_source_map(ctx_w, ctx);
   }
```
(Note `css` will be freed automatically by buffer handle)
(Same goes for source_map string)
(Requires updating all affected tests)
(Requires updating CLI: lib/render.js)
(Requires updating readme)

Thanks for not taking no for an answer! :+1:

**Comment by am11** [comment]
This is addressed by 65619ba via #729.
